devlib/utils/android.py

```diff
@@ -704,16 +635,11 @@ def wait_for(self, regexp, timeout=30):
 
         # Found some matches, return them
         # Also return if thread not running
-        if len(res) > 0 or not self.is_alive():
+        if len(res) > 0:
             return res
 
-        # Did not find any match, wait for one to pop up
-        self._sought = regexp
-        found = self._match_found.wait(timeout)
-        self._match_found.clear()
-        self._sought = None
-
-        if found:
-            return [self._found]
-        else:
+        try:
+            self._logcat.expect(regexp, timeout=timeout)
+            return [self._logcat.match.groups(0)]
```

Comment:
Hm that's a change of behaviour with the current code, but I guess it goes with how pexpect works.
By returning `match.groups(0)`, you can't pass regexps with named sections or several matching groups.

Unless I'm wrong on how pexpect returns regexp matches, it means you have different return behaviours depending if you immediately found a match (return `self.search(regexp), which is a list of lines)` or if you had to wait to find a match (return `[self._logcat.match.groups(0)]`) which is a list of matches.

Reply:
Hmm, not certain if I understand you - are you saying that this should instead return `[self._logcat.match.group(0)]` (`group` not `groups`, so it returns the full match)? If so then yes I think you are right - I probably just typo'd' this.

Reply:
What I meant is that we shouldn't be returning the match in this case, but the whole line that caused the match to happen. That's what `search()` does:
```
return [line for line in self.get_log() if re.match(regexp, line)]
```

Reply:
OK, after a bit of fiddling I discovered that [pexpect's regex matching is a little unusual](https://pexpect.readthedocs.io/en/stable/overview.html#find-the-end-of-line-cr-lf-conventions) so I've just set this to return the result of `search`. A little inefficient but it's on the host so I don't care, and it's an easy way to get consistency.

Reply:
Fair enough. Since `search` is just a one-liner, maybe we could just save the index of the last line in the log before we wait for a match, and then do a search from this index forward ?

Reply:
OK, I'll do that

Reply:
OK that's done